src/shared/dissect-image.c

```diff
@@ -1270,27 +1321,60 @@ static int verity_partition(
 #if HAVE_CRYPT_ACTIVATE_BY_SIGNED_KEY
-                if (root_hash_sig)
-                        r = crypt_activate_by_signed_key(cd, name, root_hash, root_hash_size, root_hash_sig, root_hash_sig_size, CRYPT_ACTIVATE_READONLY);
-                else {
-                        _cleanup_free_ char *hash_sig = NULL;
-                        size_t hash_sig_size;
-
-                        r = read_full_file_full(AT_FDCWD, root_hash_sig_path, 0, &hash_sig, &hash_sig_size);
-                        if (r < 0)
-                                return r;
-
-                        r = crypt_activate_by_signed_key(cd, name, root_hash, root_hash_size, hash_sig, hash_sig_size, CRYPT_ACTIVATE_READONLY);
-                }
+                        r = crypt_activate_by_signed_key(cd, name, root_hash, root_hash_size, root_hash_sig ?: hash_sig_from_file, root_hash_sig_size, CRYPT_ACTIVATE_READONLY);
 #else
-                r = log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "activation of verity device with signature requested, but not supported by cryptsetup due to missing crypt_activate_by_signed_key()");
+                        r = log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "activation of verity device with signature requested, but not supported by cryptsetup due to missing crypt_activate_by_signed_key()");
 #endif
-        } else
-                r = crypt_activate_by_volume_key(cd, name, root_hash, root_hash_size, CRYPT_ACTIVATE_READONLY);
-        if (r < 0)
-                return r;
+                } else
+                        r = crypt_activate_by_volume_key(cd, name, root_hash, root_hash_size, CRYPT_ACTIVATE_READONLY);
+                if (!IN_SET(r, 0, -EEXIST, -ENODEV, -EINVAL))
+                        return r;
+                /* libdevmapper can return EINVAL when the device is already in the activation stage.
+                 * There's no way to distinguish this situation from a genuine error due to invalid
+                 * parameters, so immediately fallback to activating the device with a unique name.
+                 * Improvements in libcrypsetup can ensure this never happens: https://gitlab.com/cryptsetup/cryptsetup/-/merge_requests/96 */
+                if (r == -EINVAL && FLAGS_SET(flags, DISSECT_IMAGE_VERITY_SHARE))
+                        return verity_partition(m, v, root_hash, root_hash_size, verity_data, NULL, root_hash_sig ?: hash_sig_from_file, root_hash_sig_size, flags & ~DISSECT_IMAGE_VERITY_SHARE, d);
+                if (r == -EEXIST) {
+                        if (!restore_deferred_remove){
+                                /* To avoid races, disable automatic removal on umount while setting up the new device. Restore it on failure. */
+                                r = dm_deferred_remove_cancel(name);
+                                if (r < 0)
+                                        return log_error_errno(r, "Disabling automated deferred removal for verity device %s failed: %m", node);
```

Comment:
downgrade log level here and everywhere...

In this file, logging above LOG_DEBUG should really be done only in functions such as `dissect_image_and_warn()` or `dissected_image_decrypt_interactively()`, (i.e. those which have in the name that they are going to be noise or interact diretcly with the user) but all library-like functions should only do debug logging, following our usual coding style. (of course, if htere's a function that is only ever called by an `xyz_and_warn()` or `xyz_interactively()` then it's fine for it to be noisy too, and log above LOG_DEBUG.